drgn_tools/slabinfo.py

```diff
@@ -0,0 +1,310 @@
+    Collect freelist data from kmem_cache_cpu and
+    return the count of free objects
+    """
+    cpu_free_set_rc = slub_get_cpu_freelist(prog, slab_cache, cpu_freelist)
+    return len(cpu_free_set_rc)
+
+
+def slub_partial_free(
+    prog: drgn.Program, slab_cache: drgn.Object, slab: drgn.Object
+) -> int:
+    """
+    Collect partial list from kmem_cache_cpu
+    """
+    cpu_partial = slab
+    pss_cnt = 0
+    partial_free = partial_objects = partial_inuse = 0
+
+    while cpu_partial:
+        page = cast("struct page *", cpu_partial)
+
+        try:
+            partial_objects = page.objects.value_()
+        except (AttributeError, drgn.ObjectAbsentError):
+            partial_objects = 0
+            return 0
+
+        try:
+            partial_inuse = page.inuse.value_()
+        except (AttributeError, drgn.ObjectAbsentError):
+            partial_inuse = 0
+            return 0
+
+        partial_free += partial_objects - partial_inuse
+
+        try:
+            cpu_partial = page.next
+        except (AttributeError, drgn.ObjectAbsentError):
+            cpu_partial = 0
+        pss_cnt += 1
+
+    return partial_free
+
+
+def kmem_cache_slub_info(
+    prog: drgn.Program, slab: drgn.Object, nrnodeid: int
+) -> None:
+    """
+    Collect the per-slab data.
+    Traverse though all the CPUs to collect the slab data.
+    """
+    total_slabs = objects = free_objects = total_inuse = 0
+
+    for cpuid in for_each_online_cpu(prog):
+        try:
+            per_cpu_slab = per_cpu_ptr(slab.cpu_slab, cpuid)
+        except (AttributeError, drgn.ObjectAbsentError):
+            continue
+        else:
+            cpu_freelist = per_cpu_slab.freelist
+            cpu_slab_ptr = per_cpu_slab.page
+            cpu_partial = per_cpu_slab.partial
+
+            if not cpu_slab_ptr:
+                continue
+
+        try:
+            page_inuse = cpu_slab_ptr.inuse.value_()
+        except (AttributeError, drgn.ObjectAbsentError):
+            page_inuse = 0
+
+        try:
+            objects = cpu_slab_ptr.objects.value_()
+        except (AttributeError, drgn.ObjectAbsentError):
+            objects = 0
+            continue
+        if objects < 0:
+            objects = 0
+
+        free_objects += objects - page_inuse
+        cpu_free_objects = slub_get_cpu_freelist_cnt(prog, slab, cpu_freelist)
+        free_objects += cpu_free_objects
+        objects = 0
+        page_inuse = 0
+        partial_frees = slub_partial_free(prog, slab, cpu_partial)
+        free_objects += partial_frees
+        total_slabs += 1
+
+    collect_node_info(
+        prog, slab, nrnodeid, total_slabs, free_objects, total_inuse
+    )
+
+
+def get_slab_info(prog: drgn.Program) -> None:
+    """
+    Print the slab headers and initiate the slab collection
+    process for all slabs
+    """
+    output.append(
+        ["CACHE", "OBJSIZE", "ALLOCATED", "TOTAL", "SLABS", "SSIZE", "NAME"]
+    )
+    nrnodeid = prog["nr_node_ids"].value_()
```

Comment:
It might make sense to skip this, not provide the `nrnodeid` parameter, and just let the `collect_node_info()` function use [for_each_online_node](https://drgn.readthedocs.io/en/latest/helpers.html#drgn.helpers.linux.nodemask.for_each_online_node) to iterate over each relevant node.